File: fsharp/BronKerbosch.Test/VertexSetTest.fs
module BronKerbosch.Test.VertexSet

open NUnit.Framework
open BronKerbosch

let nix = VertexSet.empty
let one = VertexSet.singleton (Vertex 1)
let two: VertexSet = [ 1; 2 ] |> Seq.map Verticise.it |> VertexSet.ofSeq

let six: VertexSet = [ 0; 1; 2; 3; 4; 5 ] |> Seq.map Verticise.it |> VertexSet.ofSeq

[<Test>]
let is_disjoint () =
    Assert.That(VertexSet.is_disjoint nix one)
    Assert.That(VertexSet.is_disjoint one nix)
    Assert.That(VertexSet.is_disjoint nix two)
    Assert.That(VertexSet.is_disjoint two nix)
    Assert.That(VertexSet.is_disjoint nix six)
    Assert.That(VertexSet.is_disjoint six nix)
    Assert.That(not (VertexSet.is_disjoint one two))
    Assert.That(not (VertexSet.is_disjoint two one))
    Assert.That(not (VertexSet.is_disjoint one six))
    Assert.That(not (VertexSet.is_disjoint six one))
    Assert.That(not (VertexSet.is_disjoint six two))
    Assert.That(not (VertexSet.is_disjoint two six))
    Assert.That(not (VertexSet.is_disjoint one one))
    Assert.That(not (VertexSet.is_disjoint two two))
    Assert.That(not (VertexSet.is_disjoint six six))

[<Test>]
let intersection () =
    Assert.That(VertexSet.intersect nix one, Is.Empty)
    Assert.That(VertexSet.intersect one nix, Is.Empty)
    Assert.That(VertexSet.intersect nix two, Is.Empty)
    Assert.That(VertexSet.intersect two nix, Is.Empty)
    Assert.That(VertexSet.intersect nix six, Is.Empty)
    Assert.That(VertexSet.intersect six nix, Is.Empty)
    Assert.That(VertexSet.intersect one two, Is.EqualTo<VertexSet> one)
    Assert.That(VertexSet.intersect two one, Is.EqualTo<VertexSet> one)
    Assert.That(VertexSet.intersect one six, Is.EqualTo<VertexSet> one)
    Assert.That(VertexSet.intersect six one, Is.EqualTo<VertexSet> one)
    Assert.That(VertexSet.intersect two six, Is.EqualTo<VertexSet> two)
    Assert.That(VertexSet.intersect six two, Is.EqualTo<VertexSet> two)
    Assert.That(VertexSet.intersect one one, Is.EqualTo<VertexSet> one)
    Assert.That(VertexSet.intersect two two, Is.EqualTo<VertexSet> two)
    Assert.That(VertexSet.intersect six six, Is.EqualTo<VertexSet> six)

[<Test>]
let overlap () =
    Assert.That(VertexSet.overlap nix one, Is.EqualTo 0)
    Assert.That(VertexSet.overlap one nix, Is.EqualTo 0)
    Assert.That(VertexSet.overlap nix two, Is.EqualTo 0)
    Assert.That(VertexSet.overlap two nix, Is.EqualTo 0)
    Assert.That(VertexSet.overlap nix six, Is.EqualTo 0)
    Assert.That(VertexSet.overlap six nix, Is.EqualTo 0)
    Assert.That(VertexSet.overlap one two, Is.EqualTo 1)
    Assert.That(VertexSet.overlap two one, Is.EqualTo 1)
    Assert.That(VertexSet.overlap one six, Is.EqualTo 1)
    Assert.That(VertexSet.overlap six one, Is.EqualTo 1)
    Assert.That(VertexSet.overlap two six, Is.EqualTo 2)
    Assert.That(VertexSet.overlap six two, Is.EqualTo 2)
    Assert.That(VertexSet.overlap one one, Is.EqualTo 1)
    Assert.That(VertexSet.overlap two two, Is.EqualTo 2)
    Assert.That(VertexSet.overlap six six, Is.EqualTo 6)

[<Test>]
let difference () =
    Assert.That(VertexSet.difference nix one, Is.Empty)
    Assert.That(VertexSet.difference nix two, Is.Empty)
    Assert.That(VertexSet.difference nix six, Is.Empty)
    Assert.That(VertexSet.difference one one, Is.Empty)
    Assert.That(VertexSet.difference one two, Is.Empty)
    Assert.That(VertexSet.difference one six, Is.Empty)
    Assert.That(VertexSet.difference two two, Is.Empty)
    Assert.That(VertexSet.difference two six, Is.Empty)
    Assert.That(VertexSet.difference six six, Is.Empty)
    Assert.That(VertexSet.difference one nix, Is.EqualTo<VertexSet> one)
    Assert.That(VertexSet.difference two nix, Is.EqualTo<VertexSet> two)
    Assert.That(VertexSet.difference six nix, Is.EqualTo<VertexSet> six)
    Assert.That(VertexSet.difference two one, Is.EqualTo<VertexSet>(VertexSet.singleton (Vertex 2)))

    Assert.That(
        VertexSet.difference six one,
        Is.EqualTo<VertexSet>([ 0; 2; 3; 4; 5 ] |> Seq.map Verticise.it |> VertexSet.ofSeq)
    )

    Assert.That(
        VertexSet.difference six two,
        Is.EqualTo<VertexSet>([ 0; 3; 4; 5 ] |> Seq.map Verticise.it |> VertexSet.ofSeq)
    )

[<Test>]
let pop_arbitrary_1 () =
    let x, rest = VertexSet.pop_arbitrary one
    Assert.That(x.Value, Is.EqualTo(Vertex 1))
    Assert.That(rest, Is.Empty)
    let x, rest = VertexSet.pop_arbitrary rest
    Assert.That(x, Is.Null)
    Assert.That(rest, Is.Empty)

[<Test>]
let pop_arbitrary_2 () =
    let x, rest = VertexSet.pop_arbitrary two
    let y, rest = VertexSet.pop_arbitrary rest
    Assert.That(min x.Value y.Value, Is.EqualTo(Vertex 1))
    Assert.That(max x.Value y.Value, Is.EqualTo(Vertex 2))
    Assert.That(rest, Is.Empty)
    let x, rest = VertexSet.pop_arbitrary rest
    Assert.That(x, Is.Null)
    Assert.That(rest, Is.Empty)

File: csharp/BronKerbosch.Test/VertexSetTest.cs
using NUnit.Framework;
using System;
using System.Collections.Generic;

namespace BronKerbosch.Test
{
    public class VertexSetHashTest : VertexSetTestTemplate<HashSet<Vertex>, HashSetMgr> { }
    public class VertexSetSortedTest : VertexSetTestTemplate<SortedSet<Vertex>, SortedSetMgr> { }

    public class VertexSetTestTemplate<TVertexSet, TVertexSetMgr>
        where TVertexSet : ISet<Vertex>
        where TVertexSetMgr : IVertexSetMgr<TVertexSet>
    {
        [Test]
        public void PopArbitrary1()
        {
            TVertexSet one = TVertexSetMgr.From([Vertex.Nth(1)]);
            var x = one.PopArbitrary().Index();
            Assert.That(x, Is.EqualTo(1));
            Assert.That(one, Is.Empty);
        }

        [Test]
        public void PopArbitrary2()
        {
            TVertexSet two = TVertexSetMgr.From([Vertex.Nth(1), Vertex.Nth(2)]);
            var x = two.PopArbitrary().Index();
            var y = two.PopArbitrary().Index();
            Assert.That(Math.Min(x, y), Is.EqualTo(1));
            Assert.That(Math.Max(x, y), Is.EqualTo(2));
            Assert.That(two, Is.Empty);
        }

        [Test]
        public void Overlaps()
        {
            var empty = TVertexSetMgr.Empty();
            var one = TVertexSetMgr.From([Vertex.Nth(1)]);
            var two = TVertexSetMgr.From([Vertex.Nth(1), Vertex.Nth(2)]);
            var six = TVertexSetMgr.From([Vertex.Nth(0), Vertex.Nth(1), Vertex.Nth(2),
                                          Vertex.Nth(3), Vertex.Nth(4), Vertex.Nth(5)]);
            Assert.That(!TVertexSetMgr.Overlaps(empty, one));
            Assert.That(!TVertexSetMgr.Overlaps(one, empty));
            Assert.That(!TVertexSetMgr.Overlaps(empty, two));
            Assert.That(!TVertexSetMgr.Overlaps(two, empty));
            Assert.That(!TVertexSetMgr.Overlaps(empty, six));
            Assert.That(!TVertexSetMgr.Overlaps(six, empty));
            Assert.That(TVertexSetMgr.Overlaps(one, two));
            Assert.That(TVertexSetMgr.Overlaps(two, one));
            Assert.That(TVertexSetMgr.Overlaps(one, six));
            Assert.That(TVertexSetMgr.Overlaps(six, one));
            Assert.That(TVertexSetMgr.Overlaps(two, six));
            Assert.That(TVertexSetMgr.Overlaps(six, two));
            Assert.That(TVertexSetMgr.Overlaps(one, one));
            Assert.That(TVertexSetMgr.Overlaps(two, two));
            Assert.That(TVertexSetMgr.Overlaps(six, six));
        }

        [Test]
        public void Intersection()
        {
            var empty = TVertexSetMgr.Empty();
            var one = TVertexSetMgr.From([Vertex.Nth(1)]);
            var two = TVertexSetMgr.From([Vertex.Nth(1), Vertex.Nth(2)]);
            var six = TVertexSetMgr.From([Vertex.Nth(0), Vertex.Nth(1), Vertex.Nth(2),
                                          Vertex.Nth(3), Vertex.Nth(4), Vertex.Nth(5)]);
            Assert.That(TVertexSetMgr.Intersection(empty, one).SetEquals(empty));
            Assert.That(TVertexSetMgr.Intersection(one, empty).SetEquals(empty));
            Assert.That(TVertexSetMgr.Intersection(empty, two).SetEquals(empty));
            Assert.That(TVertexSetMgr.Intersection(two, empty).SetEquals(empty));
            Assert.That(TVertexSetMgr.Intersection(empty, six).SetEquals(empty));
            Assert.That(TVertexSetMgr.Intersection(six, empty).SetEquals(empty));
            Assert.That(TVertexSetMgr.Intersection(one, two).SetEquals(one));
            Assert.That(TVertexSetMgr.Intersection(two, one).SetEquals(one));
            Assert.That(TVertexSetMgr.Intersection(one, six).SetEquals(one));
            Assert.That(TVertexSetMgr.Intersection(six, one).SetEquals(one));
            Assert.That(TVertexSetMgr.Intersection(two, six).SetEquals(two));
            Assert.That(TVertexSetMgr.Intersection(six, two).SetEquals(two));
            Assert.That(TVertexSetMgr.Intersection(one, one).SetEquals(one));
            Assert.That(TVertexSetMgr.Intersection(two, two).SetEquals(two));
            Assert.That(TVertexSetMgr.Intersection(six, six).SetEquals(six));
        }

        [Test]
        public void IntersectCount()
        {
            var empty = TVertexSetMgr.Empty();
            var one = TVertexSetMgr.From([Vertex.Nth(1)]);
            var two = TVertexSetMgr.From([Vertex.Nth(1), Vertex.Nth(2)]);
            var six = TVertexSetMgr.From([Vertex.Nth(0), Vertex.Nth(1), Vertex.Nth(2),
                                          Vertex.Nth(3), Vertex.Nth(4), Vertex.Nth(5)]);
            Assert.That(TVertexSetMgr.IntersectionSize(empty, one).Equals(0));
            Assert.That(TVertexSetMgr.IntersectionSize(one, empty).Equals(0));
            Assert.That(TVertexSetMgr.IntersectionSize(empty, two).Equals(0));
            Assert.That(TVertexSetMgr.IntersectionSize(two, empty).Equals(0));
            Assert.That(TVertexSetMgr.IntersectionSize(empty, six).Equals(0));
            Assert.That(TVertexSetMgr.IntersectionSize(six, empty).Equals(0));
            Assert.That(TVertexSetMgr.IntersectionSize(one, two).Equals(1));
            Assert.That(TVertexSetMgr.IntersectionSize(two, one).Equals(1));
            Assert.That(TVertexSetMgr.IntersectionSize(one, six).Equals(1));
            Assert.That(TVertexSetMgr.IntersectionSize(six, one).Equals(1));
            Assert.That(TVertexSetMgr.IntersectionSize(two, six).Equals(2));
            Assert.That(TVertexSetMgr.IntersectionSize(six, two).Equals(2));
            Assert.That(TVertexSetMgr.IntersectionSize(one, one).Equals(1));
            Assert.That(TVertexSetMgr.IntersectionSize(two, two).Equals(2));
            Assert.That(TVertexSetMgr.IntersectionSize(six, six).Equals(6));
        }

        [Test]
        public void Difference()
        {
            var empty = TVertexSetMgr.Empty();
            var one = TVertexSetMgr.From([Vertex.Nth(1)]);
            var two = TVertexSetMgr.From([Vertex.Nth(1), Vertex.Nth(2)]);
            var six = TVertexSetMgr.From([Vertex.Nth(0), Vertex.Nth(1), Vertex.Nth(2),
                                          Vertex.Nth(3), Vertex.Nth(4), Vertex.Nth(5)]);
            Assert.That(TVertexSetMgr.Difference(empty, one).SetEquals(empty));
            Assert.That(TVertexSetMgr.Difference(empty, two).SetEquals(empty));
            Assert.That(TVertexSetMgr.Difference(empty, six).SetEquals(empty));
            Assert.That(TVertexSetMgr.Difference(one, one).SetEquals(empty));
            Assert.That(TVertexSetMgr.Difference(one, two).SetEquals(empty));
            Assert.That(TVertexSetMgr.Difference(one, six).SetEquals(empty));
            Assert.That(TVertexSetMgr.Difference(two, two).SetEquals(empty));
            Assert.That(TVertexSetMgr.Difference(two, six).SetEquals(empty));
            Assert.That(TVertexSetMgr.Difference(six, six).SetEquals(empty));
            Assert.That(TVertexSetMgr.Difference(one, empty).SetEquals(one));
            Assert.That(TVertexSetMgr.Difference(two, empty).SetEquals(two));
            Assert.That(TVertexSetMgr.Difference(six, empty).SetEquals(six));
            Assert.That(TVertexSetMgr.Difference(two, one).SetEquals(TVertexSetMgr.From([Vertex.Nth(2)])));
            Assert.That(TVertexSetMgr.Difference(six, one).SetEquals(TVertexSetMgr.From([Vertex.Nth(0), Vertex.Nth(2), Vertex.Nth(3), Vertex.Nth(4), Vertex.Nth(5)])));
            Assert.That(TVertexSetMgr.Difference(six, two).SetEquals(TVertexSetMgr.From([Vertex.Nth(0), Vertex.Nth(3), Vertex.Nth(4), Vertex.Nth(5)])));
        }
    }
}
